core/src/main/java/com/dylibso/mcpx4j/core/Mcpx.java

```diff
@@ -56,7 +63,34 @@ public Mcpx build() {
             if (httpClientAdapter == null) {
                 httpClientAdapter = new JdkHttpClientAdapter();
             }
-            return new Mcpx(apiKey, baseUrl, jsonDecoder, httpClientAdapter, config);
+            return new Mcpx(apiKey, baseUrl, profileSlug, jsonDecoder, httpClientAdapter, config);
+        }
+
+        /**
+         * @param profile is a list of components for a slug. It may be empty (then it defaults to `~/default`),
+         *                it may contain one single value `"default"`.
+         *                Otherwise, it must contain two components, the first being a username
+         *                and the second being a profile name.
+         *                Neither the profile name nor the username may contain a slash.
+         *                It is also allowed to pass the values `"~", "default"`, yielding `~/default`.
+         */
+        static String profileIdToSlug(String... profile) {
+            if (profile.length == 0) {
+                return "~/default";
+            }
+            if (profile.length == 1) {
+                if (profile[0].contains("/")) {
+                    throw new IllegalArgumentException("Invalid profile path: " + Arrays.toString(profile));
+                }
+                return "~/" + profile[0];
+            }
+            if (profile.length > 2) {
+                throw new IllegalArgumentException("Invalid profile path: " + Arrays.toString(profile));
+            }
+            if (profile[0].contains("/") ||  profile[1].contains("/")) {
+                throw new IllegalArgumentException("Invalid profile path: " + Arrays.toString(profile));
+            }
+            return profile[0] + "/" + profile[1];
```

Comment:
```suggestion
            if (profile.length == 0) {
                return "~/default";
            }
            if (profile.length == 1) {
                return "~/" + URLEncoder.encode(profile[0], "UTF-8");
            }
            if (profile.length > 2) {
                throw new IllegalArgumentException("Invalid profile path: " + Arrays.toString(profile));
            }
            return URLEncoder.encode(profile[0], "UTF-8") + "/" + URLEncoder.encode(profile[1], "UTF-8");

```

Sorry for this last curve-ball – the components of a profile _can_ contain url-encoded variables. E.g., I can have a profile named `chrisdickinson//!\nerd alert/!\`, which would have a slug of `chrisdickinson/%2F!\nerdalert%2F!\`. If I wanted to use this in `mcpx4j`, I'd probably try passing `.withProfile("~", "/!\\nerd%20alert/!\\")` (with the expectation that I'm handing the literal components over for mcpx4j to build a url-safe slug.)

Reply:
(Important caveat here: you know Java users much better than I – so if you think they're more likely to pass pre-url-encoded profile components here, I'm totally okay skipping this change – whichever you think is best!)

Reply:
no worries! and no I think it's safer to urlencode!

Reply:
I didn't even notice the suggestion, thanks :P